### PR TITLE
 Handle python types for query parameters

### PR DESCRIPTION
**Pull request by rushton**
Prior to this commit, the only python types supported as query parameters were str and int. In this commit the types are expanded to bool, float, bytes, datetime, list, dict, and UUID.

### CODE REVIEW

**Comment by rushton** [comment]
does not yet work for older versions of Presto

**Comment by rushton** [comment]
the fix fails with presto version >=329.

File: presto/dbapi.py [hunk 1]
```diff
@@ -281,17 +282,44 @@ def _prepare_statement(self, operation, statement_name):
...
+            return "true" if param else "false"
+        if isinstance(param, float):
+            return "DOUBLE '%s'" % param
+        if isinstance(param, str):
+            return ("'%s'" % param.replace("'", "''"))
+        if isinstance(param, bytes):
+            return "X'%s'" % param.hex()
+        if isinstance(param, datetime.datetime):
+            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z").rstrip(" ")
+            return "TIMESTAMP '%s'" % re.sub(r'\.([0-9]{3})([0-9]{3})', r'.\1', datetime_str)
```
Comment (rushton) [inline presto/dbapi.py:312]: prior to presto 335 the timestamp format only supports 1/100th of a second in the timestamp format, thus the last 3 digits of the subseconds are truncated (effectively dividing by 1000).

File: presto/dbapi.py [hunk 2]
```diff
@@ -281,17 +282,46 @@ def _prepare_statement(self, operation, statement_name):
...
+        if isinstance(param, list):
+            return "ARRAY[%s]" % ','.join(map(self._format_prepared_param, param))
+        if isinstance(param, dict):
+            keys = self._format_prepared_param(list(param.keys()))
+            values = self._format_prepared_param(list(param.values()))
+            return "MAP(%s, %s)" % (keys, values)
+        if isinstance(param, uuid.UUID):
+            return "UUID '%s'" % str(param)
+
+        return str(param)
```
Comment (rushton) [inline presto/dbapi.py:322]: thought an exception might be appropriate here, rather than presto spitting out a random syntax error. Thoughts?

Comment (findepi) [reply presto/dbapi.py:322]: Agreed.

File: presto/dbapi.py (hunk 1 above)
Comment (findepi) [reply presto/dbapi.py:312]: we can have a legacy mode for dealing with older versions, but we should be correct by default (no implicit truncation nor rounding) for the new versions

doing truncation via regular expression is very clever, but i am thinking we could perhaps round using some methods on `param`, or by providing desired output format to `strftime` directly.

File: integration_tests/fixtures.py [hunk 3]
```diff
@@ -29,7 +29,7 @@
 logger = presto.logging.get_logger(__name__)
 
 
-PRESTO_VERSION = os.environ.get("PRESTO_VERSION") or "329"
+PRESTO_VERSION = os.environ.get("PRESTO_VERSION") or "330"
```
Comment (findepi) [inline integration_tests/fixtures.py:32]: 344 is the latest

File: integration_tests/test_dbapi.py [hunk 4]
```diff
@@ -87,22 +90,45 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
     Tests string parameters in prepared statements
     """
     cur = presto_connection.cursor()
+    expected_uuid = uuid.uuid1()
     # test a simple string parameter and
     # a parameter with a quote in it
     cur.execute(
-        "select ?",
-        params=("six'",),
+        "SELECT ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?",
```
Comment (findepi) [inline integration_tests/test_dbapi.py:102]: This is efficient (one query, many assertions), but this does not "scale" when you add more test cases (the list of question marks will grow and grow), or more assertions -- for example we will want to verify not only how `SELECT ?` looks like back on our side, but also what Presto engine things of the type of `?` (eg `SELECT typeof(?)`)

Thus, I would rather go with separate query + assertions for every value tested.
This could be similar to what we do with JDBC tests.
(I am linking an example to a read path test, but i am sure you get what i mean)
https://github.com/prestosql/presto/blob/cbcba450481feb3d7b92fb862dd6692b6a921f7e/presto-jdbc/src/test/java/io/prestosql/jdbc/BaseTestJdbcResultSet.java#L188-L202

File: presto/dbapi.py [hunk 5]
```diff
@@ -298,13 +299,27 @@ def _format_prepared_param(self, param):
         Formats parameters to be passed in an
         EXECUTE statement.
         """
+        if isinstance(param, bool):
```
Comment (findepi) [inline presto/dbapi.py:302]: add
```
if param is None:
    return "NULL"
```

as the first case

File: presto/dbapi.py [hunk 6]
```diff
@@ -298,13 +299,27 @@ def _format_prepared_param(self, param):
...
+            return "DOUBLE '%s'" % param
         if isinstance(param, str):
             return ("'%s'" % param.replace("'", "''"))
-        else:
-            # TODO: param could have many other types which should
-            #       be covered including unicode, maps, lists, etc.
-            #       https://github.com/prestosql/presto-python-client/issues/45
-            return str(param)
+        if isinstance(param, bytes):
+            return "X'%s'" % param.hex()
```
Comment (findepi) [inline presto/dbapi.py:309]: can you please add support for unicode as well?
for now this can be as simple as the `str` case.
in the future we can go with `U&'...'` unicode escapes, but maybe we do not actually need to

File: presto/dbapi.py [hunk 7]
```diff
@@ -298,13 +299,27 @@ def _format_prepared_param(self, param):
...
+        if isinstance(param, datetime.datetime):
+            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z").rstrip(" ")
+            return "TIMESTAMP '%s'" % re.sub(r'\.([0-9]{3})([0-9]{3})', r'.\1', datetime_str)
+        if isinstance(param, list):
+            return "ARRAY[%s]" % ','.join(map(self._format_prepared_param, param))
+        if isinstance(param, dict):
+            keys = self._format_prepared_param(list(param.keys()))
+            values = self._format_prepared_param(list(param.values()))
```
Comment (findepi) [inline presto/dbapi.py:317]: Can `param.keys()` and `param.values()` ordering go out of sync? can it depend on `dict` subclass?
if possible, this could be a safe guard:
```
entries = list(param. items())
keys = self._format_prepared_param([k for (k, v) in entries])
values = self._format_prepared_param([v for (k, v) in entries])
```

File: integration_tests/test_dbapi.py [hunk 8]
```diff
@@ -87,22 +90,45 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
+        params=(
+            "six'",
+            datetime(2020,1,1,0,0,0),
+            datetime(2020,1,1,0,0,0,tzinfo=pytz.utc),
```
Comment (findepi) [inline integration_tests/test_dbapi.py:106]: that's good test case to have

how would i construct "timestamp with time zone" equivalent with just standard library?
should we test for this as well?

File: integration_tests/test_dbapi.py [hunk 9]
```diff
@@ -87,22 +90,45 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
+    expected_uuid = uuid.uuid1()
     # test a simple string parameter and
     # a parameter with a quote in it
     cur.execute(
-        "select ?",
-        params=("six'",),
+        "SELECT ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?",
+        params=(
+            "six'",
+            datetime(2020,1,1,0,0,0),
```
Comment (findepi) [inline integration_tests/test_dbapi.py:105]: nit: while we do not have a formal code style estabilished here, i personally would prefer a space after a comma

File: presto/dbapi.py (hunk 5 above)
Comment (rushton) [reply presto/dbapi.py:304]: done, thanks!

File: presto/dbapi.py (hunk 2 above)
Comment (rushton) [reply presto/dbapi.py]: added, used `NotSupportedError` as the exception. Let me know if that is the correct one to use  here.

File: integration_tests/fixtures.py (hunk 3 above)
Comment (rushton) [reply integration_tests/fixtures.py]: updated

File: presto/dbapi.py (hunk 7 above)
Comment (rushton) [reply presto/dbapi.py]: Correct, they can be out of order, good catch.  I ended up using the keys to build the list of values.

File: integration_tests/test_dbapi.py (hunk 8 above)
Comment (rushton) [reply integration_tests/test_dbapi.py]: This does output a `timestamp with time zone` type from presto. I've added an assertion to verify that is the case.

File: integration_tests/test_dbapi.py (hunk 9 above)
Comment (rushton) [reply integration_tests/test_dbapi.py]: updated, thanks!

File: presto/dbapi.py (hunk 1 above)
Comment (rushton) [reply presto/dbapi.py:316]: just to clarify, should we only support the later versions here and forego truncating the subseconds? or round depending on how the user constructed the `datetime`?

File: integration_tests/test_dbapi.py (hunk 4 above)
Comment (rushton) [reply integration_tests/test_dbapi.py]: broke out into multiple tests, thanks for the tip!

File: presto/dbapi.py (hunk 6 above)
Comment (rushton) [reply presto/dbapi.py:313]: did some reading, in python 3 all `str` objects are unicode, thus should we always apply the `U&` prefix? also, any suggestions for doing unicode detection that is compatible with both  python2 and python3?

File: integration_tests/test_dbapi.py [hunk 10]
```diff
@@ -87,22 +90,97 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
+        params=([[1, 2, 3],[4,5,6]],))
+    rows = cur.fetchall()
+
+    assert rows[0][0] == [[1, 2, 3],[4,5,6]]
+
+
+def test_dict_query_params(presto_connection):
+    cur = presto_connection.cursor()
+
+    cur.execute("SELECT ?", params=({"foo": "bar"},))
```
Comment (findepi) [inline integration_tests/test_dbapi.py:150]: please add 
```
# TODO validate the type on the Presto side (e.g. `typeof(?)`)
```

File: integration_tests/test_dbapi.py [hunk 11]
```diff
@@ -87,22 +90,97 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
-    assert len(rows) == 1
-    assert rows[0][0] == "six'"
+    assert rows[0][0] == "2020-01-01 00:00:00.000 UTC"
+    assert cur.description[0][1] == "timestamp with time zone"
+
+
+def test_array_query_params(presto_connection):
+    cur = presto_connection.cursor()
+
+    cur.execute("SELECT ?", params=([1, 2, 3],))
```
Comment (findepi) [inline integration_tests/test_dbapi.py:134]: please add 
```
# TODO validate the type on the Presto side (e.g. `typeof(?)`)
```

File: integration_tests/test_dbapi.py [hunk 12]
```diff
@@ -87,22 +90,97 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
+
+    cur.execute("SELECT ?", params=(False,))
+    rows = cur.fetchall()
+
+    assert rows[0][0] == False
+
+
+def test_float_query_params(presto_connection):
+    cur = presto_connection.cursor()
+    cur.execute("SELECT ?", params=(1.1,))
```
Comment (findepi) [inline integration_tests/test_dbapi.py:172]: add test cases with NaN, +inf and -inf

File: integration_tests/test_dbapi.py [hunk 13]
```diff
@@ -87,22 +90,97 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
+    rows = cur.fetchall()
+
+    assert rows[0][0] == 1.1
+
+
+def test_int_query_params(presto_connection):
+    cur = presto_connection.cursor()
+    cur.execute("SELECT ?", params=(3,))
```
Comment (findepi) [inline integration_tests/test_dbapi.py:180]: add a test case with negative number

add a test case with number not fitting in 32-bit in, eg Long.MAX_VALUE ( 9,223,372,036,854,775,807 )

File: presto/dbapi.py [hunk 14]
```diff
@@ -298,13 +299,35 @@ def _format_prepared_param(self, param):
         Formats parameters to be passed in an
         EXECUTE statement.
         """
+        if param is None:
+            return "NULL"
+        if isinstance(param, bool):
+            return "true" if param else "false"
+        if isinstance(param, int):
+            return "%d" % param
```
Comment (findepi) [inline presto/dbapi.py:307]: add
```
# TODO represent numbers exceeding 64-bit (BIGINT) as DECIMAL
```

File: presto/dbapi.py (hunk 6 above)
Comment (findepi) [reply presto/dbapi.py:313]:
```
isinstance(param, type(u''))
```

or, perhaps more elegant way
```
try:
    from builtins import unicode
except ImportError:
    unicode = str
```

File: presto/dbapi.py (hunk 1 above)
Comment (findepi) [reply presto/dbapi.py:316]: I do not see a need to support old versions here.

we can always revisit if there is enough demand

File: presto/dbapi.py [hunk 15]
```diff
@@ -298,13 +299,35 @@ def _format_prepared_param(self, param):
...
             return ("'%s'" % param.replace("'", "''"))
-        else:
-            # TODO: param could have many other types which should
-            #       be covered including unicode, maps, lists, etc.
-            #       https://github.com/prestosql/presto-python-client/issues/45
-            return str(param)
+        if isinstance(param, bytes):
+            return "X'%s'" % param.hex()
+        if isinstance(param, datetime.datetime):
+            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z").rstrip(" ")
```
Comment (findepi) [inline presto/dbapi.py:315]:
```
datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z")
# strip trailing whitespace if param has no zone
datetime_str = datetime_str.rstrip(" ")
```

File: presto/dbapi.py [hunk 16]
```diff
@@ -298,13 +299,35 @@ def _format_prepared_param(self, param):
...
+        if isinstance(param, dict):
+            keys = list(param.keys())
+            values = [param[key] for key in keys]
+
+            return "MAP(%s, %s)" % (
+                self._format_prepared_param(keys),
+                self._format_prepared_param(values)
+            )
+        if isinstance(param, uuid.UUID):
+            return "UUID '%s'" % str(param)
```
Comment (findepi) [inline presto/dbapi.py:328]: is `str` redundant here?

**Review by findepi — APPROVED** [review]
thanks, minor comments

File: presto/dbapi.py (hunk 6 above)
Comment (findepi) [reply presto/dbapi.py:313]: > should we always apply the U& prefix

let's not use U& for now. just handling `unicode` _type_ is enough for this PR.

as a follow up we can add some unicode test cases (with ascii, non-ascii, non-BMP, etc.)

File: integration_tests/test_dbapi.py (hunk 10 above)
Comment (rushton) [reply integration_tests/test_dbapi.py:157]: added the test case for this

File: integration_tests/test_dbapi.py (hunk 11 above)
Comment (rushton) [reply integration_tests/test_dbapi.py:135]: added the test case for this

File: integration_tests/test_dbapi.py (hunk 12 above)
Comment (rushton) [reply integration_tests/test_dbapi.py:184]: added tests and implementation for these cases, but they all fail on the results side. Using `infinity()` and `nan()` functions in presto convert to a `str` in the python client. I'm not sure where this is happening in the code, but willing to fix it if you have some insight. For now I've marked the tests as skip with an explanation.

File: integration_tests/test_dbapi.py (hunk 13 above)
Comment (rushton) [reply integration_tests/test_dbapi.py:214]: added

File: presto/dbapi.py (hunk 14 above)
Comment (rushton) [reply presto/dbapi.py:309]: added

File: presto/dbapi.py (hunk 15 above)
Comment (rushton) [reply presto/dbapi.py:323]: done

File: presto/dbapi.py (hunk 1 above)
Comment (rushton) [reply presto/dbapi.py]: removed truncation of subseconds

File: presto/dbapi.py (hunk 16 above)
Comment (rushton) [reply presto/dbapi.py]: yes, fixed

File: integration_tests/test_dbapi.py (hunk 12 above)
Comment (findepi) [reply integration_tests/test_dbapi.py:184]: > For now I've marked the tests as skip with an explanation.

Thanks, please file an issue to track this and let's thus unblock this PR

File: integration_tests/test_dbapi.py [hunk 17]
```diff
@@ -87,24 +91,141 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
+    assert rows[0][0] == "2020-01-01 00:00:00.000"
+
+    cur.execute(
+            "SELECT ?", 
+            params=(datetime(2020, 1, 1, 0, 0, 0, tzinfo=pytz.utc),)
+            )
+    rows = cur.fetchall()
+
+    assert rows[0][0] == "2020-01-01 00:00:00.000 UTC"
+    assert cur.description[0][1] == "timestamp with time zone"
```
Comment (findepi) [inline integration_tests/test_dbapi.py:130]: Curious, why `rows` is `[0][0]` and description is `[0][1]`?

File: integration_tests/test_dbapi.py [hunk 18]
```diff
@@ -87,24 +91,141 @@ def test_select_query_result_iteration_statement_params(presto_connection):
...
+
+    assert rows[0][0] == False
+
+
+def test_float_query_param(presto_connection):
+    cur = presto_connection.cursor()
+    cur.execute("SELECT ?", params=(1.1,))
+    rows = cur.fetchall()
+
+    assert rows[0][0] == 1.1
```
Comment (findepi) [inline integration_tests/test_dbapi.py:187]: let's have `cur.description`  assertion for floats too 
(eg to "document" floats end up as being `double` in Presto)

File: presto/dbapi.py [hunk 19]
```diff
@@ -298,13 +300,45 @@ def _format_prepared_param(self, param):
...
+            # TODO represent numbers exceeding 64-bit (BIGINT) as DECIMAL
+            return "%d" % param
+        if isinstance(param, float) and param == float("+inf"):
+            return "infinity()"
+        if isinstance(param, float) and param == float("-inf"):
+            return "-infinity()"
+        if isinstance(param, float) and math.isnan(param):
+            return "nan()"
+        if isinstance(param, float):
+            return "DOUBLE '%s'" % param
```
Comment (findepi) [inline presto/dbapi.py:317]: let's group this by type:
```
if isinstance(param, float):
    if param == float("+inf"):
       ....
    ....
    return "DOUBLE '%s'" % param
```

File: presto/dbapi.py (hunk 6 above)
Comment (findepi) [reply presto/dbapi.py:321]: OK, let's keep unicode out of this PR. https://github.com/prestosql/presto-python-client/issues/53

File: presto/dbapi.py [hunk 20]
```diff
@@ -298,13 +300,45 @@ def _format_prepared_param(self, param):
...
             return ("'%s'" % param.replace("'", "''"))
-        else:
-            # TODO: param could have many other types which should
-            #       be covered including unicode, maps, lists, etc.
-            #       https://github.com/prestosql/presto-python-client/issues/45
-            return str(param)
+        if isinstance(param, bytes):
+            return "X'%s'" % param.hex()
+        if isinstance(param, datetime.datetime):
+            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z").rstrip(" ")
```
Comment (findepi) [inline presto/dbapi.py:323]: the `rstrip` is now on the next line

```suggestion
            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z")
```

File: presto/dbapi.py [hunk 21]
```diff
@@ -298,13 +300,45 @@ def _format_prepared_param(self, param):
...
+            # strip trailing whitespace if param has no zone
+            datetime_str = datetime_str.rstrip(" ")
+
```
Comment (findepi) [inline presto/dbapi.py:326]: remove this line

(i would rather keep empty line after each `return` than within the "then" block)

File: presto/dbapi.py [hunk 22]
```diff
@@ -298,13 +300,45 @@ def _format_prepared_param(self, param):
...
+            return "TIMESTAMP '%s'" % datetime_str
+        if isinstance(param, list):
+            return "ARRAY[%s]" % ','.join(map(self._format_prepared_param, param))
+        if isinstance(param, dict):
+            keys = list(param.keys())
+            values = [param[key] for key in keys]
+
```
Comment (findepi) [inline presto/dbapi.py:333]: remove this line

**Review by findepi — APPROVED** [review]

File: integration_tests/test_dbapi.py (hunk 17 above)
Comment (rushton) [reply integration_tests/test_dbapi.py:130]: `description` is a list of tuples of the form: `[("<output_col_name>", "<output_col_type>"),...]`

Comment (findepi) [reply integration_tests/test_dbapi.py:130]: obviously... i think i now understand the feeling you had seeing my question 😉

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
many thanks!

File: integration_tests/test_dbapi.py (hunk 12 above)
Comment (rushton) [reply integration_tests/test_dbapi.py:184]: created: https://github.com/prestosql/presto-python-client/issues/54 thanks!
